File: go/src/pm/keys/keys.go
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

// Package keys provides utilities for generating package manager signing keys
package keys

import (
	"crypto/rand"
	"io"

	"golang.org/x/crypto/ed25519"
)

// tests override this for deterministic output
var randSource = rand.Reader

// Gen generates a new keypair and writes the private key to the given writer.
// The generated key contains a private and public key suitable for use with EdDSA.
func Gen(w io.Writer) error {
	_, key, err := ed25519.GenerateKey(randSource)
	if err != nil {
		return err
	}

	_, err = w.Write(key)
	return err
}

File: src/connectivity/bluetooth/core/bt-gap/src/store/keys.rs
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    anyhow::Error,
    fuchsia_bluetooth::{
        error::Error as BtError,
        types::{Address, PeerId},
    },
};

pub const BONDING_DATA_PREFIX: &'static str = "bonding-data:";
pub const HOST_DATA_PREFIX: &'static str = "host-data:";

pub fn bonding_data_key(device_id: PeerId) -> String {
    format!("{}{}", BONDING_DATA_PREFIX, device_id)
}

pub fn host_data_key(host_address: &Address) -> String {
    format!("{}{}", HOST_DATA_PREFIX, host_address.to_string())
}

pub fn host_id_from_key(key: &str) -> Result<String, Error> {
    if key.len() <= HOST_DATA_PREFIX.len() {
        return Err(BtError::new(&format!("malformed host data key: {}", key)).into());
    }
    Ok(key[HOST_DATA_PREFIX.len()..].to_string())
}
